src/java.desktop/share/classes/sun/java2d/cmm/lcms/LCMSProfile.java

```diff
@@ -25,69 +25,63 @@
 
 package sun.java2d.cmm.lcms;
 
-import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.StampedLock;
 
 import sun.java2d.cmm.Profile;
 
 final class LCMSProfile extends Profile {
-    private final TagCache tagCache;
 
     private final Object disposerReferent;
+    private final Map<Integer, byte[]> tags = new ConcurrentHashMap<>();
+    private final StampedLock lock = new StampedLock();
 
     LCMSProfile(long ptr, Object ref) {
         super(ptr);
-
         disposerReferent = ref;
-
-        tagCache = new TagCache(this);
     }
 
     long getLcmsPtr() {
-        return this.getNativePtr();
-    }
-
-    TagData getTag(int sig) {
-        return tagCache.getTag(sig);
+        return getNativePtr();
     }
 
-    void clearTagCache() {
-        tagCache.clear();
-    }
-
-    private static final class TagCache  {
-        private final LCMSProfile profile;
-        private final HashMap<Integer, TagData> tags = new HashMap<>();
-
-        private TagCache(LCMSProfile p) {
-            profile = p;
+    byte[] getProfileData() {
+        long stamp = lock.readLock();
+        try {
+            return LCMS.getProfileDataNative(getNativePtr());
+        } finally {
+            lock.unlockRead(stamp);
         }
+    }
 
-        private TagData getTag(int sig) {
-            TagData t = tags.get(sig);
-            if (t == null) {
-                byte[] tagData = LCMS.getTagNative(profile.getNativePtr(), sig);
-                if (tagData != null) {
-                    t = new TagData(tagData);
-                    tags.put(sig, t);
-                }
-            }
+    byte[] getTag(int sig) {
+        byte[] t = tags.get(sig);
+        if (t != null) {
             return t;
         }
-
-        private void clear() {
-            tags.clear();
+        long stamp = lock.readLock();
+        try {
+            return tags.computeIfAbsent(sig, (key) -> {
+                return LCMS.getTagNative(getNativePtr(), key);
+            });
+        } finally {
+            lock.unlockRead(stamp);
         }
```

Comment:
The comments about the change in this class and the changes in the "getTag()" method.
1. I have removed the TagData and TagCache classes and move the tag cache to the profile class directly.
2. I have moved synchronization logic from the LCMS.java to this class, so now we can use more "granular" synchronizations.

The logic behind this change:
1. The StampedLock guards the access to the native lcms code, so when we change the tags via "LCMS.setTagDataNative" nobody will call "LCMS.getTagNative" and "LCMS.getProfileDataNative". I tried the "ReentrantReadWriteLock", but it is a little bit slower.
2. The cache itself is maintained by the ConcurrentHashMap, and the usage of "byte[] t = tags.get(sig);" w/o synchronization is a key for the performance.

Note that it is possible to use "tags.computeIfAbsent" instead of "tags.get(sig)" and take care of "native access" synchronization in the "LCMS.getTagNative", but unfortuntly for some workload the "get()" is x10 times faster than computeIfAbsent() if the key is already exists in the map(common situation for the cache).